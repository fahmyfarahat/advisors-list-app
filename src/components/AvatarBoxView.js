import React from 'react';
import PropTypes from 'prop-types';
import {
  Grid,
  Badge, 
  Box,
  Avatar,
  Typography
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';

const OnlineBadgeStyles = withStyles(theme => ({
  badge: {
    backgroundColor: '#44b700',
    boxShadow: `0 0 0 2px ${theme.palette.background.paper}`,
    '&::after': {
      position: 'absolute',
      top: 0,
      left: 0,
      width: '100%',
      height: '100%',
      borderRadius: '50%',
      animation: '$ripple 1.5s infinite ease-in-out',
      border: '1px solid #44b700',
      content: '""',
    },
  },
  '@keyframes ripple': {
    '0%': {
      transform: 'scale(.8)',
      opacity: 1,
    },
    '100%': {
      transform: 'scale(2.4)',
      opacity: 0,
    },
  },
}))(Badge);

const AvatarBoxView = ({ src, name, status }) => (
  <Grid container spacing={1}>
    <Grid item>
      <Box>
        <OnlineBadgeStyles
          invisible={status === 'offline' ? true : false}
          overlap="circle"
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'right',
          }}
          variant="dot"
        >
          <Avatar alt={name} src={src} />
        </OnlineBadgeStyles>
      </Box>
    </Grid>
    <Grid item>
      <Typography variant="subtitle1">{name}</Typography>
    </Grid>
  </Grid>
);

AvatarBoxView.propTypes = {
  status: PropTypes.oneOf(['online', 'offline']).isRequired,
  src: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired
};

export default AvatarBoxView;