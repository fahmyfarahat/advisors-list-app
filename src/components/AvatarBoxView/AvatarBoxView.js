import React from 'react';
import PropTypes from 'prop-types';
import withStyles from '@material-ui/core/styles/withStyles';
import {
  Grid,
  Badge, 
  Box,
  Avatar,
  Typography
} from '@material-ui/core';

const OnlineBadgeStyles = withStyles(theme => ({
  badge: {
    backgroundColor: '#44b700',//TODO:: use theme colors
    boxShadow: `0 0 0 2px ${theme.palette.background.paper}`,
    '&::after': {
      position: 'absolute',
      top: 0,
      left: 0,
      width: '100%',
      height: '100%',
      borderRadius: '50%',
      animation: '$ripple 1.5s infinite ease-in-out',
      border: '1px solid #44b700',//TODO:: use theme colors
      content: '""',
    },
  },
  '@keyframes ripple': {
    '0%': {
      transform: 'scale(.8)',
      opacity: 1,
    },
    '100%': {
      transform: 'scale(2.4)',
      opacity: 0,
    },
  },
}))(Badge);

const AvatarBoxView = ({ src, name, status }) => (
  <Grid container spacing={1}>
    <Grid item>
      <Box>
        <OnlineBadgeStyles
          invisible={status === 'offline' ? true : false}
          overlap="circle"
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'right',
          }}
          variant="dot"
        >
          <Avatar alt={name} src={src} />
        </OnlineBadgeStyles>
      </Box>
    </Grid>
    <Grid item>
      <Typography variant="subtitle1">{name}</Typography>
    </Grid>
  </Grid>
);

AvatarBoxView.propTypes = {
  status: PropTypes.oneOf(['online', 'offline']).isRequired,
  src: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired
};

export default AvatarBoxView;