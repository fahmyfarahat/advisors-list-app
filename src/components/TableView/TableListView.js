import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { 
  Table,
  TableBody,
  TableCell,
  TableRow,
  Paper,
  Link,
  LinearProgress
} from "@material-ui/core";
import Rating from "@material-ui/lab/Rating";
import { tableSort, getSorting, tableFilter } from "../../utilities";
import TableToolbarView from "./TableToolbarView";
import TableHeadView from "./TableHeadView";
import AvatarBoxView from "../AvatarBoxView/AvatarBoxView";

const TableListView = ({ loading, advisorData, classes }) => {
  const [order, setOrder] = useState("asc");
  const [orderBy, setOrderBy] = useState("");
  const [advisorRows, setAdvisorRows] = useState([]);
  const [filters, setFilters] = React.useState({
    status: {
      online: false,
      offline: false
    },
    language: {
      german: false,
      english: false,
      spanish: false,
      italian: false
    }
  });
  
  useEffect(() => {
    setAdvisorRows([...advisorData]);
  }, [advisorData]);

  const handleRequestSort = (event, property) => {
    const isDesc = orderBy === property && order === "desc";
    setOrder(isDesc ? "asc" : "desc");
    setOrderBy(property);
  };
  
  const handleChange = (key, supKey) => event => {
    setFilters({
      ...filters, 
      [key] : {
        ...filters[key],
        [supKey]: event.target.checked
      }
    });
  };

  return (
    <Paper className={classes.paper}>
      <TableToolbarView 
        filters={filters} 
        onChange={handleChange}
      />
      <div className={classes.tableWrapper}>
        <Table
          className={classes.table}
          size="medium"
        >
          <TableHeadView
            order={order}
            orderBy={orderBy}
            onRequestSort={handleRequestSort}
          />
          <TableBody>
            {tableSort(tableFilter(advisorRows, filters), getSorting(order, orderBy))
              .map(row => {
                return (
                  <TableRow
                    hover
                    role="checkbox"
                    key={row.name}
                  >
                    <TableCell component="th">
                      <AvatarBoxView status={row.status} src={row.avatar} name={row.name} />
                    </TableCell>
                    <TableCell align="right">
                      <span className={classes.status}>{row.status}</span>
                    </TableCell>
                    <TableCell align="left">{row.language}</TableCell>
                    <TableCell align="left"> 
                      <Rating value={row.rating} precision={0.5} readOnly />
                      <Link>
                        {row.reviews} Reviews
                      </Link>
                    </TableCell>
                  </TableRow>
                );
              })}
          </TableBody>
        </Table>
        { loading && (<LinearProgress />)}
      </div>
      {!advisorRows.length && !loading && null}
    </Paper>
  );
};

TableListView.propTypes = {
  loading: PropTypes.bool.isRequired, 
  advisorData: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired
};

export default TableListView;