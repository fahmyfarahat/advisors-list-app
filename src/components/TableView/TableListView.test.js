import React from "react";
import { shallow, mount } from "enzyme";
import TableListView from "./TableListView";
import { LinearProgress } from "@material-ui/core";
import toJson from "enzyme-to-json";
import AvatarBoxView from "../AvatarBoxView/AvatarBoxView";

describe("TableListView - general ", () => {
  let props;
  beforeEach(()=> {
    props = {
      loading: false, 
      advisorData: [], 
      classes: {}
    };
  });

  it("should match Snapshot", () => {
    const wrapper = shallow(<TableListView {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render <LinearProgress/> if loading is true", () => {
    props.loading = true;
    const wrapper = shallow(<TableListView {...props} />);
    expect(wrapper.find(LinearProgress)).toHaveLength(1);
  })

  it("should render <TableRow/>", () => {
    props.advisorData = [{
      "id": 1,
      "name": "Loraine Mueller",
      "avatar": "url",
      "status": "offline",
      "language": "german",
      "rating": 4,
      "reviews": 4308
    }];
    const wrapper = mount(<TableListView {...props} />);
    expect(wrapper.find(AvatarBoxView)).toHaveLength(1);
  });
})
