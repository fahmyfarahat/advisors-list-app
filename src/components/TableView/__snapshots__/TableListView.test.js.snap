// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableListView - general  should match Snapshot 1`] = `
<WithStyles(ForwardRef(Paper))>
  <TableToolbarView
    filters={
      Object {
        "language": Object {
          "english": false,
          "german": false,
          "italian": false,
          "spanish": false,
        },
        "status": Object {
          "offline": false,
          "online": false,
        },
      }
    }
    onChange={[Function]}
  />
  <div>
    <WithStyles(ForwardRef(Table))
      size="medium"
    >
      <TableHeadView
        onRequestSort={[Function]}
        order="asc"
        orderBy=""
      />
      <WithStyles(ForwardRef(TableBody)) />
    </WithStyles(ForwardRef(Table))>
  </div>
</WithStyles(ForwardRef(Paper))>
`;
